gateway-server/src/main/java/org/apache/knox/gateway/services/token/impl/JDBCTokenStateService.java

```diff
@@ -0,0 +1,180 @@
+    try {
+      boolean removed = tokenDatabase.removeTokens(tokenIds);
+      if (removed) {
+        log.removedTokensFromDatabase(tokenIds.size());
+      } else {
+        log.failedToRemoveTokensFromDatabase(tokenIds.size());
+      }
+    } catch (SQLException e) {
+      log.errorRemovingTokensFromDatabase(tokenIds.size(), e.getMessage(), e);
+    }
+    super.removeTokens(tokenIds);
+  }
+
+  @Override
+  public long getTokenExpiration(String tokenId, boolean validate) throws UnknownTokenException {
+    try {
+      // check the in-memory cache, then
+      return super.getTokenExpiration(tokenId, validate);
+    } catch (UnknownTokenException e) {
+      // It's not in memory
+    }
+
+    long expiration = 0;
+    try {
+      expiration = tokenDatabase.getTokenExpiration(tokenId);
+      log.fetchedExpirationFromDatabase(Tokens.getTokenIDDisplayText(tokenId), expiration);
+    } catch (SQLException e) {
+      log.errorFetchingExpirationFromDatabase(Tokens.getTokenIDDisplayText(tokenId), e.getMessage(), e);
+    }
+    return expiration;
+  }
+
+  @Override
+  protected void updateExpiration(String tokenId, long expiration) {
+    // Update in-memory
+    super.updateExpiration(tokenId, expiration);
+
+    try {
+      final boolean updated = tokenDatabase.updateExpiration(tokenId, expiration);
+      if (updated) {
+        log.updatedExpirationInDatabase(Tokens.getTokenIDDisplayText(tokenId), expiration);
+      } else {
+        log.failedToUpdateExpirationInDatabase(Tokens.getTokenIDDisplayText(tokenId), expiration);
+      }
+    } catch (SQLException e) {
+      log.errorUpdatingExpirationInDatabase(Tokens.getTokenIDDisplayText(tokenId), e.getMessage(), e);
+    }
+  }
+
+  @Override
+  protected long getMaxLifetime(String tokenId) {
+    long maxLifetime = super.getMaxLifetime(tokenId);
+
+    // If there is no result from the in-memory collection, proceed to check the Database
+    if (maxLifetime < 1L) {
+      try {
+        maxLifetime = tokenDatabase.getMaxLifetime(tokenId);
+        log.fetchedMaxLifetimeFromDatabase(Tokens.getTokenIDDisplayText(tokenId), maxLifetime);
+      } catch (SQLException e) {
+        log.errorFetchingMaxLifetimeFromDatabase(Tokens.getTokenIDDisplayText(tokenId), e.getMessage(), e);
+      }
+    }
+    return maxLifetime;
+  }
+
+  @Override
+  protected boolean isUnknown(String tokenId) {
+    boolean isUnknown = super.isUnknown(tokenId);
+
+    // If it's not in the cache, then check in the Database
+    if (isUnknown) {
+      try {
+        isUnknown = tokenDatabase.getMaxLifetime(tokenId) < 0;
+      } catch (SQLException e) {
+        log.errorFetchingMaxLifetimeFromDatabase(Tokens.getTokenIDDisplayText(tokenId), e.getMessage(), e);
+      }
+    }
+    return isUnknown;
+  }
+
+  @Override
+  protected List<String> getTokenIds() {
+    List<String> tokenIds = new LinkedList<>();
+    try {
+      tokenIds = tokenDatabase.getTokenIds();
```

Comment:
This seems like an expensive operation and can potentially block a lot of CPU

Reply:
This is a simple `SELECT * FROM` query, nothing special happens here. Even if this was the case we need to get them from the Database to support HA.
My two cents:
- in case of `AliasBasedTokenStateService` there is a separate thread that reads all relevant aliases from the keystore into the memory -> this method does not need to be overridden. This is strictly non-HA.
- in case of `ZKTokenStateService` the listener mechanism makes sure the given Knox node collects all aliases into the in-memory -> this method does not need to be overridden.

Please note this method invoked by the reaper thread to evict expired tokens. This means we could do a better job here by adding a  `WHERE` clause to filter records that are expired already (or close enough to be expired).

Reply:
My concern is the SELECT clause blowing up, atleast we should use LIMIT clause 
ie
`SELECT * FROM tbl LIMIT 5,10;  # Retrieve rows 6-15`

Reply:
With the new implementation this is not a problem anymore (I did not even overwrite this method)